### PR TITLE
Enable tablet

### PR DESCRIPTION
**Pull request by dy**
Fixes https://github.com/plotly/plotly.js/issues/280

### CODE REVIEW

**Comment by etpinard** [comment]
@nicolaskruchten would you mind checking out https://codepen.io/etpinard/pen/wEOyJG on your iPad?

Thanks!

**Comment by nicolaskruchten** [comment]
I’ll get the iPad checked soon but FYI it doesn’t work on iPhone 7

**Comment by etpinard** [comment]
Maybe try https://codepen.io/etpinard/full/wEOyJG/

**Comment by etpinard** [comment]
compare to https://codepen.io/etpinard/full/GXexQM/ with `plotly-latest`.

**Comment by etpinard** [comment]
@dy things still don't work on iPad.

Here's a screenshot of Chrome on the "latest generation" iPad:

image

Let me know if you need a refresher on how to use Browserstack.

**Comment by nicolaskruchten** [comment]
Confirmed on a physical iPad

**Comment by dy** [comment]
Ok, now that works. Thanks guys for the feedback. That took [a while](https://github.com/juliangruber/is-mobile/pull/10) to make correct API.

**Comment by nicolaskruchten** [comment]
Lmk if there’s a link I can test! :)

**Comment by dy** [comment]
@nicolaskruchten the same link https://codepen.io/etpinard/pen/wEOyJG should work, unless cached
image

**Comment by nicolaskruchten** [comment]
Ok so webgl works on both iPad and iPhone, nice job! But... on iPad the surface plot doesn’t show up initially... only if I drag-rotate it. It sometimes flickers on and then disappears leaving just the title and colorbar

**Comment by nicolaskruchten** [comment]
So to be clear, the detection bug is fixed, but likely we will get additional iPad bug reports once this is down streamed :)

**Comment by etpinard** [comment]
> on iPad the surface plot doesn’t show up initially... only if I drag-rotate it

Right, I noticed the same issue on browserstack. 

--------

Is it worth merging and releasing this as part of this week's `v1.42.0`? Or should we wait for a fix for that flickers issue?

**Comment by nicolaskruchten** [comment]
Your call: is it worse to have a message saying "webgl not supported by your browser" even though it is, or to have a partial figure rendered?

**Comment by dy** [comment]
Hm that seems to be gl-plot3d specific issue. I may look into it.

**Comment by etpinard** [comment]
> Your call: is it worse to have a message saying "webgl not supported by your browser" even though it is, or to have a partial figure rendered?

Ok. I'd vote for fixing this "well" once and for all. @dy can you try to look at this next?

**Comment by dy** [comment]
Ok, should be fixed now (b62dc8f fixes #11). Updatet build, the same link should work on iPads now https://codepen.io/etpinard/pen/wEOyJG

**Comment by nicolaskruchten** [comment]
Is the above code pen still the one I should check?

**Comment by nicolaskruchten** [comment]
https://codepen.io/etpinard/pen/wEOyJG works on iPad now!

**Comment by nicolaskruchten** [comment]
And iPhone! Would be good to do more testing but at first glance it looks good to me :)

**Comment by nicolaskruchten** [comment]
It does flicker badly once but at least after the flicker it appears instead of being blank

**Comment by nicolaskruchten** [comment]
Ah... not always. Sometimes it still seems to be blank

**Comment by nicolaskruchten** [comment]
Huh... it’s much more reliable on both devices in “full” mode than in “pen” mode so could be a code pen artifact?

**Comment by dy** [comment]
Oh it does flicker indeed. Let me check if plotly build is right, gl-plot3d seemed to have that fixed.
That still flickers with more complex objects.

**Comment by dy** [comment]
Since removing `dirty` flag and `gl.clear()` calls (forcing scene to draw every frame) keeps on flickering, the only explanation is that `requestAnimationFrame` is separate from WebGL frames in iOS in some way. Likely some scene frames take longer to draw than actual gl frames, which effects in skipped blank frames. Keeping `preserveDrawingBuffer: true` for now.

(the pen https://codepen.io/etpinard/pen/wEOyJG seems to be fixed now @nicolaskruchten)

**Comment by nicolaskruchten** [comment]
That’s much better on both devices!

**Comment by nicolaskruchten** [comment]
While I’m picking at nits though, the image initially flashes on at an odd camera angle then snaps into place. Tapping “zoom” or “pan” locks it back into that odd angle on both devices for some reason

**Comment by dy** [comment]
@nicolaskruchten that seems to be not only ipad issue, I have it on laptop too (clicking zoom/pan).

**Comment by nicolaskruchten** [comment]
Ok, great! I’m happy, so long as these changes don’t destabilize other platforms/OSes

File: scene.js [hunk 1]
```diff
@@ -80,7 +80,8 @@ function createScene(options) {
     gl = getContext(canvas,
       options.glOptions || {
         premultipliedAlpha: true,
-        antialias: true
+        antialias: true,
+        preserveDrawingBuffer: true
```
Comment (etpinard) [inline scene.js]: Could this be `preserveDrawingBuffer: isMobile`, to not deteriorate performance?

**Comment by etpinard** [comment]
> Ok, great! I’m happy, so long as these changes don’t destabilize other platforms/OSes

All tests are passing. I didn't notice any differences in behaviour while testing out our gl3d mocks. 

Waiting on a reply for https://github.com/gl-vis/gl-plot3d/pull/11#discussion_r223723506 by @dy, but this appears like a solid improvement :1st_place_medal:

File: scene.js (hunk 1 above)
Comment (dy) [reply scene.js]: @etpinard that does not affect performance in any way, in fact by avoiding redundant redraws it is possible to improve it. I may have a look at refactoring. Fixed.

**Comment by etpinard** [comment]
Thanks @dy !

Merging this in. This will be part of plotly.js `v1.42.0`

**Pull request merged**